### PR TITLE
[GPU] improve primitive impl caching mechanism with new unified key

### PR DESCRIPTION
**Pull request by ahnyoung-paul**
### Details:
 - *Apply new impl key using new hash function in kernel params*
 - *Filter duplicated task before entering async compilation queue*
 - *gpu functional test will be uploaded*
 - *performance test and accuracy test*

### Tickets:
 - *95115*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp [hunk 1]
```diff
@@ -59,20 +59,42 @@ class wait_for_events_impl : public primitive_impl {
     }
 
     void update_dispatch_data(const kernel_impl_params& impl_param) override { }
+
+    template<typename PType>
+    static size_t get_impl_key(const typed_program_node<PType>& arg, const kernel_impl_params& impl_param) {
+        size_t seed = 0;
+        auto& id = impl_param.desc->id;
+        for (size_t i = 0; i < id.size(); i++) {
+            seed = hash_combine(seed, id[i]);
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp:68]: Why you add primitive name and id to the hash? The initial intention was to allow reuse of impl withing single model if all the parameters + shapes are same.

File: src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp [hunk 2]
```diff
@@ -99,6 +98,18 @@ struct typed_primitive_impl_ocl : public typed_primitive_impl<PType> {
         return make_unique<ImplType>(arg, best_kernel);
     }
 
+    template<typename ImplType>
+    static size_t get_impl_key(const typed_program_node<PType>& /*arg*/, const kernel_impl_params& impl_param) {
+        auto kernel_params = ImplType::get_kernel_params(impl_param);
+        auto params = kernel_params.first;
+#if 1
+        return params.hash();
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp:106]: I'd expect hash to be a part of primitive, not kernel parameters. If you keep it in the kernel parameters, you'll need to implement hash function for all impl types of the same primitive, while having hash for primitive + adding impl type + shapes could be generic solution. Moreover, hash for primitive can be computed once on graph load, so in runtime you'll just need to compute hash for runtime params, and no need to convert parameters to kernel selector format (which may be expensive too)

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp:106]: Thank you for your comments. I will update it.

File: src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp (hunk 1 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp:68]: Indeed. It should not contain the id.  @ahnyoung-paul Was the experiment based on this implementation using id in hash? If so, it is also weird how the kernel could be reused across the layers.

File: src/plugins/intel_gpu/src/graph/impls/ocl/arg_max_min.cpp [hunk 3]
```diff
@@ -60,22 +60,22 @@ struct arg_max_min_impl : typed_primitive_impl_ocl<arg_max_min> {
...
-    static std::unique_ptr<primitive_impl> create(const arg_max_min_node& arg, const kernel_impl_params& impl_param) {
+    static kernel_params_t get_kernel_params(const kernel_impl_params& impl_param) {
         const auto& primitive = impl_param.typed_desc<arg_max_min>();
         const auto& axis = primitive->axis;
         const auto& top_k = primitive->top_k;
         const auto& mode = primitive->mode;
         const auto& sort_type = primitive->sort;
         const auto& values_first = primitive->values_first;
-        const auto& outputs_num = arg.get_output_nums();  // second output passed as input for TOP_K layer
+        const auto& outputs_num = (primitive->input_size() == 3 ? 2 : primitive->output_size());  // second output passed as input for TOP_K layer
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/impls/ocl/arg_max_min.cpp:70]: Why this is changed?

File: src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp (hunk 1 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp:68]: @ahnyoung-paul  Seems that this is not as we aligned last time. Isnt this not the final impl?

File: src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp (hunk 2 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/impls/ocl/primitive_base.hpp:106]: >> primitive + adding impl type + shape
Is it enough for the key? Even though the shape and impl type is same, there should be other parameters to be checked, like kernel parameters (padding, fused ops, etc..)

File: src/plugins/intel_gpu/src/graph/impls/ocl/arg_max_min.cpp (hunk 3 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/ocl/arg_max_min.cpp:70]: because I want to remove arg_max_min_node arg from get_kernel_params's input parameters.

File: src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp (hunk 1 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp:68]: It should be removed. I forget to remove it.

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/impls/common/wait_for_events.cpp]: Removed it.

File: src/plugins/intel_gpu/src/graph/compilation_context.cpp [hunk 4]
```diff
@@ -28,8 +30,15 @@ class CompilationContext : public ICompilationContext {
...
-    void push_task(ICompilationContext::Task&& task) override {
-        _queue.push(task);
+    void push_task(size_t key, ICompilationContext::Task&& task) override {
+        std::lock_guard<std::mutex> lock(_mutex);
+        auto iter = _key_map.find(key);
+        if (iter == _key_map.end()) {   // not found
+            auto insert_it = _data_list.insert(_data_list.end(), {key, task});
+            _key_map.insert({key, insert_it});
+        } else {    // found
+            _data_list.splice(_data_list.begin(), _data_list, iter->second);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/compilation_context.cpp:40]: why is this needed?

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/compilation_context.cpp:40]: It has better performance of first running benchmark_app to use this code when the disk cache is used (cl_cache_dir is set). because the most used kernel is built first.

Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/compilation_context.cpp:40]: Aligned offline

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/compilation_context.cpp:40]: removed it

File: src/plugins/intel_gpu/include/intel_gpu/primitives/average_unpooling.hpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/primitives/average_unpooling.hpp:3]: This file is removed in the previous commit.

File: src/plugins/intel_gpu/include/intel_gpu/runtime/layout.hpp [hunk 6]
```diff
@@ -332,6 +332,14 @@ struct padding {
         return padding{lower.sizes(), upper.sizes(), filling_value};
     }
 
+    size_t hash() const {
+        size_t seed = 0;
+        seed = cldnn::hash_combine(seed, _filling_value);
+        seed = cldnn::hash_combine(seed, _lower_size.hash());
+        seed = cldnn::hash_combine(seed, _upper_size.hash());
+        return seed;
+    }
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/runtime/layout.hpp:341]: Isnt the layout shape calculated by this hash?

File: src/plugins/intel_gpu/include/intel_gpu/runtime/layout.hpp [hunk 7]
```diff
@@ -503,6 +511,19 @@ struct layout {
...
 
+    size_t hash() const {
+        size_t seed = 0;
+        seed = hash_combine(seed, data_padding.hash());
+        seed = hash_combine(seed, format.value);
+        seed = hash_combine(seed, data_type);
+        if (!is_dynamic()) {
+            for (auto& d : get_shape()) {
+                seed = hash_combine(seed, d);
+            }
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/runtime/layout.hpp:522]: Can't  we calculate the dynamic shape layout? I think we can use the partial shape value -1. 
Even though it is dynamic shape, it may have static dimensions partially. (e.g., {1, 2, -1, -1} .. I such a case, we might distinguish. (For now we only use this hash mechanism for impl cache with static shapes, but in the future we might need to extend the usage for compilation stage too..)

File: src/plugins/intel_gpu/include/intel_gpu/runtime/tensor.hpp [hunk 8]
```diff
@@ -318,6 +318,26 @@ struct tensor {
...
+            seed = hash_combine(seed, feature[i]);
+        }
+
+        for (size_t i = 0; i < spatial.size(); ++i) {
+            seed = hash_combine(seed, spatial[i]);
+        }
+
+        for (size_t i = 0; i < group.size(); ++i) {
+            seed = hash_combine(seed, group[i]);
+        }
```
Comment (yeonbok) [inline src/plugins/intel_gpu/include/intel_gpu/runtime/tensor.hpp:337]: minor comment but can we replace the above with hash_range? :)

File: src/plugins/intel_gpu/src/graph/compilation_context.cpp [hunk 9]
```diff
@@ -10,16 +10,19 @@ namespace cldnn {
 
 class CompilationContext : public ICompilationContext {
 public:
-    using compilation_queue_t = InferenceEngine::ThreadSafeQueue<ICompilationContext::Task>;
+    using data_type = std::pair<size_t, ICompilationContext::Task>;
+    using data_list_type = std::list<data_type>;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/compilation_context.cpp:14]: Now we just use it as queue. Should we still  maintain this as list?

File: src/plugins/intel_gpu/src/graph/compilation_context.cpp [hunk 10]
```diff
@@ -40,11 +48,32 @@ class CompilationContext : public ICompilationContext {
...
+            task = front.second;
+            return true;
+        }
+        return false;
+    }
+
+    void pop_front_task() {
+        std::lock_guard<std::mutex> lock(_mutex);
+        auto front = _compile_queue.front();
+        _compile_queue_keymap.erase(front.first);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/compilation_context.cpp:65]: As I asked offline, erasing this from keymap as soon as pop, the duplicate enqueue might occur while the same task is still being compiled. 
So I think erasing should happen afther the compilation is done.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 11]
```diff
@@ -1246,4 +1228,16 @@ void primitive_inst::load(cldnn::BinaryInputBuffer& ib) {
         _intermediates_memory[i] = get_network().get_engine().allocate_memory(ibuf_layout, _allocation_type);
     }
 }
+
+size_t primitive_inst::get_layout_key(const kernel_impl_params& params) const {
+    size_t seed = _node->hash();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:1233]: Now it has more info in the hash than layout. 
Would be better to rename the function :)

File: src/plugins/intel_gpu/src/graph/compilation_context.cpp (hunk 10 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/compilation_context.cpp:65]: pop_front_task() is called after compilation is done. so it is erased and I think it is working as we expected.
Could you please check it again?

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/compilation_context.cpp:65]: Change function name from pop_front_task to erase_front_task.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp (hunk 11 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/primitive_inst.cpp:1233]: Change func name from get_layout_key to get_impl_key

File: src/plugins/intel_gpu/include/intel_gpu/primitives/average_unpooling.hpp (hunk 5 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/average_unpooling.hpp:3]: Removed it.

File: src/plugins/intel_gpu/src/graph/compilation_context.cpp (hunk 9 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/compilation_context.cpp]: Replaced to std::dequeu

File: src/plugins/intel_gpu/include/intel_gpu/runtime/layout.hpp (hunk 7 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/layout.hpp]: I changed codes to get hash value using parital_shape. thanks.

File: src/plugins/intel_gpu/include/intel_gpu/runtime/tensor.hpp (hunk 8 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/tensor.hpp]: Changed it.

File: src/plugins/intel_gpu/src/graph/compilation_context.cpp (hunk 10 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/graph/compilation_context.cpp:63]: But, shouldnt the popping from the queue itself be called right after the task began to executed? 
Otherwise, other thread can get the task from the compilation queue.
keymap should be erased after the compilation finished to prevent pushing same task. 
However removing from the queue should be done as soon as the compilation is begin.. ?

File: src/tests/functional/plugin/gpu/subgraph_tests/dynamic/dynamic_smoke_test_gen_impl_key.cpp [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        std::map<std::string, std::string> additionalConfig;
+        std::tie(inputShapes, netType, targetDevice, additionalConfig) = basicParamsSet;
+
+        init_input_shapes(inputShapes);
+        const auto inShapeShapeOf = inputDynamicShapes[0];
+        const auto inShapeElt = inputDynamicShapes[1];
+        auto params = builder::makeDynamicParams(netType, {inShapeShapeOf, inShapeElt});
+        auto paramOuts = helpers::convert2OutputVector(ngraph::helpers::castOps2Nodes<ngraph::opset3::Parameter>(params));
+
+        // Left
```
Comment (yeonbok) [inline src/tests/functional/plugin/gpu/subgraph_tests/dynamic/dynamic_smoke_test_gen_impl_key.cpp:89]: what does this mean?

Comment (ahnyoung-paul) [reply src/tests/functional/plugin/gpu/subgraph_tests/dynamic/dynamic_smoke_test_gen_impl_key.cpp]: It is redundant comment. removed it

File: src/plugins/intel_gpu/src/graph/compilation_context.cpp (hunk 10 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/compilation_context.cpp]: Update compilation queue. please review it again. thanks.

File: src/plugins/intel_gpu/src/graph/include/convolution_inst.h [hunk 13]
```diff
@@ -107,6 +107,18 @@ struct typed_program_node<convolution> : public typed_program_node_base<convolut
...
     }
 
+    size_t hash() const override {
+        if (!seed) {
+            seed = parent::hash();
+
+            seed = hash_combine(seed, transposed);
+            seed = hash_combine(seed, groups);
+            seed = hash_combine(seed, deformable_groups);
+            seed = hash_combine(seed, deformable_mode);
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/include/convolution_inst.h:117]: Same here for everything except `transposed`.

File: src/plugins/intel_gpu/src/graph/include/program_node.h [hunk 14]
```diff
@@ -494,6 +497,35 @@ struct typed_program_node_base : public program_node {
         return std::static_pointer_cast<const PType>(program_node::get_primitive());
     }
 
+    size_t hash() const override {
+        if (!seed) {
+            // hash for primitive
+            seed = get_primitive()->hash();
+
+            // hash for type
+            primitive_id type_str = get_primitive()->type_string();
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/graph/include/program_node.h:506]: Why this is not a part of primitive hash?

File: src/plugins/intel_gpu/include/intel_gpu/primitives/convolution.hpp [hunk 15]
```diff
@@ -763,6 +763,20 @@ struct convolution : public primitive_base<convolution> {
     /// @brief List of primitive ids containing compensation.
     primitive_id_arr compensation;
 
+    size_t hash() const override {
+        if (!seed) {
+            seed = hash_range(seed, pad.begin(), pad.end());
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/include/intel_gpu/primitives/convolution.hpp:768]: I guess number of inputs and outputs must be a part of hash as it may impact mode of work for some kernels (conv w/ and w/o bias, concat, etc)

File: src/tests/functional/plugin/gpu/subgraph_tests/dynamic/dynamic_smoke_test_gen_impl_key.cpp [hunk 16]
```diff
@@ -0,0 +1,189 @@
+// Copyright (C) 2018-2022 Intel Corporation
```
Comment (vladimir-paramuzov) [inline src/tests/functional/plugin/gpu/subgraph_tests/dynamic/dynamic_smoke_test_gen_impl_key.cpp:1]: nit: 2023

File: src/plugins/intel_gpu/src/graph/include/deconvolution_inst.h [hunk 17]
```diff
@@ -43,6 +43,15 @@ struct typed_program_node<deconvolution> : public typed_program_node_base<deconv
         return params;
     }
 
+    size_t hash() const override {
+        if (!seed) {
+            seed = parent::hash();
+
+            seed = hash_combine(seed, groups);
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/include/deconvolution_inst.h:50]: I see, I will remove it.

File: src/plugins/intel_gpu/src/graph/include/convolution_inst.h (hunk 13 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/include/convolution_inst.h:117]: I see, I will remove it.

File: src/tests/functional/plugin/gpu/subgraph_tests/dynamic/dynamic_smoke_test_gen_impl_key.cpp (hunk 16 above)
Comment (ahnyoung-paul) [reply src/tests/functional/plugin/gpu/subgraph_tests/dynamic/dynamic_smoke_test_gen_impl_key.cpp:1]: Fixed it.

File: src/plugins/intel_gpu/include/intel_gpu/primitives/primitive.hpp [hunk 18]
```diff
@@ -120,6 +123,7 @@ struct primitive {
     size_t num_outputs;
 
 protected:
+    mutable size_t seed = 0;
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/primitive.hpp]: Remove it from primitive. program_node already have hash variable.

File: src/plugins/intel_gpu/src/graph/include/program_node.h (hunk 14 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/include/program_node.h]: moved it.

File: src/plugins/intel_gpu/src/graph/include/convolution_inst.h (hunk 13 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/include/convolution_inst.h]: Removed it.

File: src/plugins/intel_gpu/src/graph/include/deconvolution_inst.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/include/deconvolution_inst.h]: Removed it.

File: src/plugins/intel_gpu/include/intel_gpu/primitives/convolution.hpp (hunk 15 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/convolution.hpp]: Add number of inputs / outputs to hash functions

File: src/plugins/intel_gpu/src/graph/include/convolution_inst.h [hunk 19]
```diff
@@ -107,6 +107,15 @@ struct typed_program_node<convolution> : public typed_program_node_base<convolut
         return params;
     }
 
+    size_t hash() const override {
+        if (!seed) {
+            seed = parent::hash();
+
+            seed = hash_combine(seed, transposed);
+        }
+        return seed;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/include/convolution_inst.h:116]: Can this be fixed to use primitive::hash() && hash_combine(seed, transposed) ?

**Comment by yeonbok** [comment]
LGTM except minor comment

File: src/plugins/intel_gpu/src/graph/include/convolution_inst.h (hunk 19 above)
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/include/convolution_inst.h:116]: I think it is hard to change it. because primitive::hash() is used only in primitive but this is convolution_node which is not primitive. 
If primitive::hash() is used instead of parent::hash(), we should add codes to calculate hash of activation and fused params in convolution_node::hash().

**Comment by ahnyoung-paul** [comment]
@yeonbok @vladimir-paramuzov Could you please review it again? thanks.

File: src/plugins/intel_gpu/include/intel_gpu/primitives/primitive.hpp [hunk 20]
```diff
@@ -85,6 +85,17 @@ struct primitive {
...
     virtual primitive_id type_string() const = 0;
 
+    virtual size_t hash() const {
+        size_t seed = 0;
+        // hash for type
+        primitive_id type_str = type_string();
+        for (size_t idx = 0; idx < type_str.size(); idx++) {
+            seed = hash_combine(seed, type_str[idx]);
+        }
+        seed = hash_combine(seed, num_outputs);
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/include/intel_gpu/primitives/primitive.hpp:97]: Add hash for number of inputs in primitive::hash()

File: src/plugins/intel_gpu/src/graph/graph_optimizer/compile_graph.cpp [hunk 21]
```diff
@@ -21,6 +21,7 @@ void compile_graph::run(program& p) {
     OV_ITT_SCOPED_TASK(ov::intel_gpu::itt::domains::intel_gpu_plugin, "pass::CompileGraph");
     for (auto& node : p.get_processing_order()) {
         node->set_unique_id();
+        node->hash();
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/compile_graph.cpp]: @vladimir-paramuzov, Separate program_node::hash() as your comments. Please review it again. thanks.

**Comment by ahnyoung-paul** [comment]
To fix gpu unit test failure issue, I moved location of nodes' hash calculation from compile_graph to program ctor because when optimization option is false compile_graph is not called. @vladimir-paramuzov , @yeonbok please review it again.

File: src/plugins/intel_gpu/src/graph/primitive_inst.cpp [hunk 22]
```diff
@@ -283,24 +283,6 @@ void primitive_inst::update_impl() {
         return l.transform(format::bfwzyx).to_shape();
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/primitive_inst.cpp:283]: (random spot) 
BTW shouldnt we update the key for update_weight too?

**Comment by yeonbok** [comment]
Note: update_weights is still using unique id in layout key. It seems that we can improve the update_weight perf with unified key. Lets do it in a separate work.

**Review by vladimir-paramuzov — APPROVED** [review]
Overall, LGTM. I'd recommend to add few unit tests for hash values (like create primitive - get hash - compare with hardcoded value, then create program node, calculate hash, etc) to ensure that hashing is stable.

**Comment by ahnyoung-paul** [comment]
> Overall, LGTM. I'd recommend to add few unit tests for hash values (like create primitive - get hash - compare with hardcoded value, then create program node, calculate hash, etc) to ensure that hashing is stable.

I added simple unit tests to check hash value. please review unit test tool

**Review by yeonbok — APPROVED** [review]
